documentation/guide-api-first.asciidoc

```diff
@@ -0,0 +1,144 @@
+
+In case you only want to generate your API resources once and version them afterwards in SCM, simply run the generator, outputting to some temporary location, then, copy them to `src/main/java` and you are done. Be careful if you manually modify the generated files afterwards and you want to re-generate them after an API schema update, you will lose any manual changes. 
+
+=== Implement the generated interfaces
+
+To implement the generated interfaces, we simply create an impl class - rest controller bean - that implements the interface from our `gen` package:
+
+[source, java]
+----
+package org.acme.rest.controller;
+
+import java.util.List;
+
+import com.petclinic.api.PetsApi;
+import com.petclinic.api.model.Pet;
+
+public class PetClinicController implements PetsApi {
+
+    @Override
+    public void createPets() {
+        // TODO Auto-generated method stub
+        
+    }
+
+    @Override
+    public List<Pet> listPets(Integer limit) {
+        // TODO Auto-generated method stub
+        return null;
+    }
+
+    @Override
+    public Pet showPetById(String petId) {
+        // TODO Auto-generated method stub
+        return null;
+    }
+
+}
+----
+
+And now we can invoke our API endpoint as usual: `http://localhost:8080/pets` - because `/pets` is the `@Path` annotation value in the generated `PetsApi` interface.
+
+=== Serving API docs / Swagger-UI
+
+A common requirement is that our backend API provider should also provide an endpoint with the schema or a Swagger-UI application with that schema. 
+In our example, we decided to generate the JAX-RS interface without Swagger/OpenAPI annotations, therefore the schema can not be re-constructed 1:1 from our code (missing method documentation, error handling, etc.).
+
+When having a Quarkus application and using the link:guide-openapi.asciidoc#smallrye-openapi[Smallrye OpenAPI extension], we can tell Quarkus to serve a static version of the API as our openapi schema (the same file we used to generate the interfaces and models) and to disable the auto-generating of the schema. Follow the https://quarkus.io/guides/openapi-swaggerui#openapi[Quarkus OpenAPI documentation] for more info.
+
+=== Advanced topics
+
+In some cases, we may have specific requirements or API extensions that are not supported by the existing generators. OpenAPI generator project allows us to define https://github.com/OpenAPITools/openapi-generator/tree/master/modules/openapi-generator-maven-plugin#custom-generator[custom genererator], or to extend the existing https://openapi-generator.tech/docs/templating[generator templates]. We can also selectively generate subset of the models or API endpoints, generate test code and much more. 
```

Comment:
OpenAPI generator project allows us to define **a** [custom genererator],....

We can also selectively generate (a subset OR subsets) of the models or API endpoints